### ISSUE TITLE
"KvCommunitheme"

### ISSUE THREAD

**Issue by tsujan** [tsujan/Kvantum#189]
"KvCommunitheme" -- as a fair clone of Ubuntu's new "Communitheme" -- is almost ready but, since Ubuntu designers change "Communitheme" once in a while, it can't be added yet. This is a reminder for modifying and adding it after Ubuntu decides how "Communitheme" should look at last.

EDIT: The theme is already added but needs (continuous?) updates.

**Comment by CDrummond** [comment]
So far, so good. Great work on this initial version. Some issues:
1. Yeah, buttons are due to change again!
2. Gtk does not have animation on button press
3. Entry focus should be darker on inside, and ligher on outside
4. Entry needs internal shadow when not focused
5. Kvantum's pushbuttons are getting taller when pressed? Looks odd
6. Kvantum's toolbar buttons have frey underline on hover, and orange when pressed. Gtk's do not do this. This looks more like Adapta than Communitheme. Communitheme's dark toolbar buttons are highlighted be a light rounded square. When pressed an internal shadow is added
7. View's seem to 'pulse' when focused - you get inner frame and then it disappears. Looks odd. Think KvAmbiance used to do this too.
8. Active tab title should be bold
9. Any chance you could draw tabs a little smaller. and drag grey background? I know this would be faking the Gtk look, but worth a try? Up to you, not a big deal to me.
10. Sliders have orange border on left when focused? Focused react should be about 3 pixels away from groove - but not as large as slider.
11. Menu shadow is too much.
12. Weird scrollbar on combo menus
13. With editable combos - perhaps draw a line to the left of the arrow? Might make look match a little closer?
14. Button focus rect is not centred. The bottom line looks closer to the bottom of the button than the other sides. I guess this is due to the shadow.
15. Keyboard shortcut underline is sometimes drawn as a double thickness line.
16. Unfocused windows are slightly darker
17. Widgets in unfocused windows do not react to mouse-over. I think this is wrong though, and have raised a Communitheme issue: https://github.com/ubuntu/gtk-communitheme/issues/194

**Comment by tsujan** [comment]
Thank you very much for checking it! So fast!

> Gtk does not have animation on button press

I know but Kvantum's animation is all-or-nothing. I like those animations ;)

> Entry focus should be darker on inside, and ligher on outside
> Entry needs internal shadow when not focused

Although entries have a mild internal shadow with KvCommunitheme, I think we have different versions of Communitheme and should wait for further changes.

> Kvantum's pushbuttons are getting taller when pressed? Looks odd

They don't get taller but just *look* so because the thickness of the shadow below the normal button should be included in the button height and it disappears with the toggled state. The same is true with KvAdapta, which have a thick shadow below its buttons. There's no other way with Qt.

> Kvantum's toolbar buttons have frey underline on hover, and orange when pressed. Gtk's do not do this. 

It does here. This is again about different versions of Communitheme we have.

> View's seem to 'pulse' when focused - you get inner frame and then it disappears.

That's because of animation. I won't remove it ;)

> Active tab title should be bold

Again, different versions of Communitheme... although, in Kvantum, either both active and inactive tabs have a bold text or none of them has. I may change that later though.

> Any chance you could draw tabs a little smaller. and drag grey background?

That GTK look of tabs is impossible with Qt. Smaller tabs wouldn't provide enough space for joining tabs to the tab widget elegantly.

> Sliders have orange border on left when focused? Focused react should be about 3 pixels away from groove - but not as large as slider.

That would be very ugly with some themes. IMO, the focus rect should either surround the whole slider or be on its groove frame. Both are possible with Kvantum and I may change it to the former with this theme.

> Menu shadow is too much.

OK.

> Weird scrollbar on combo menus

They're the theme scrollbars, nothing else. Do you see another kind of scrollbar there?

> With editable combos - perhaps draw a line to the left of the arrow?

It's quite possible and I tested that: very ugly, mainly because of the bottom shadow. In GTK, the bottom shadow is almost removed from combos. In Qt, non-editable combos should look like buttons, IMO.

> Button focus rect is not centred

Because of the shadows Communitheme's buttons have. As mentioned above, shadows are included in the button height. However, I may have a solution -- actually, I was thinking about it before.

> Keyboard shortcut underline is sometimes drawn as a double thickness line.

Here, they're OK. There should be something wrong with rendering on GTK DEs.

> Unfocused windows are slightly darker

Again, this should be related the GTK DE you use. Here, there's no difference -- because there's no inactive state (yet).

> Widgets in unfocused windows do not react to mouse-over.

Exactly like above. None of these happens here.

**Comment by CDrummond** [comment]
I'm always using the latest Communitheme - I update daily (multiple times a day :-) )

I agree to wait for Communitheme to settle down - I was just listing the differences. Even as it currently is, KvCommunitheme is **very** good :-) I'm just a nit-picking perfectionist :-)

Pushbuttons - I can see what you mean. But it looks very odd, and I dont see why they need to extend down. Just draw a transparent area? i.e. transparent shadow. The current solution is suboptimal. Shadow issue also affects toolbuttons.

Toolbar buttons - I have **never** seen this underline in Communitheme buttons. The nearest would be in the nautilus pathbar, or the linked buttons in the about dialogs. etc. Perhaps the orange line below is what you are mistaking as toolbar highlight?

![screenshot from 2018-03-31 18-14-37](https://user-images.githubusercontent.com/1141694/38165703-758649fa-350f-11e8-9ecd-0c46dc029a7b.png)

Communitheme does not draw dark toolbars, only headerbars. I know I said to draw the primary toolbar dark (and this _might_ still be a good idea). In the screenshot above you can see the toolbar button highlight for a light toolbar - a grey are is drawn. Only when pressed is the frame drawn. KvCommunitheme is drawing frame even on mouse over.

View pulse - please, even if you have animation this **needs** to be fixed, its distracting.

Tabs - fair enough. As I said not a major issue. But, please, bold for active tab? This setting might be useful for other themes too.

Sliders - OK, but current focus (where its on just one side of the groove) is odd. A compromise would be to draw around complete widget. This would look nicer.

Editable Combos - really? Just a slight line (not touching the frame) looks ugly? I'll take your word for it. Again, this is not a major issue.

Button focus rect - can the this not be adjust to take into account the shadow? Communitheme is surely not the only theme where a button has shadow, and hence the contents are not in the exact middle.

Keyboard shortcut underline - Yeah, probably Qt issue. Think I've seen this with Fusion.

Unfocused window - I'm using GNOME shell, but its not the WM that controls the darkness of the unfocused window. Communitheme itself explicitly makes the widgets of unfocused windows darker. https://github.com/ubuntu/gtk-communitheme/issues/212 And I've already linked the bug report about the widgets in inactive windows not reacting to mouse over. The theme itself blocks these, so should be WM/DE independent. 

As to the scrollbar in combo menus see:

![kvantum-scrollbar](https://user-images.githubusercontent.com/1141694/38165819-93e05d30-3511-11e8-8088-c46ad2388819.png)


I don't expect KvCommunitheme to be a 100% exact copy of Communitheme, as I understand Qt is not Gtk. Different widgets, and different behaviours. Still, it would be nice to be a close as possible.

**Comment by tsujan** [comment]
For now, I fixed the taller toggled button by using an old SVG trick (in https://github.com/tsujan/Kvantum/commit/786166635f31d54076cd8d06ce1c1c3e96cec0b7). I'll look into the focus rect later.

> I'm always using the latest Communitheme - I update daily (multiple times a day :-) )

That should be the cause. I don't have Ubuntu and get it from https://launchpad.net/~communitheme/+archive/ubuntu/ppa once in 4-5 days.

> Pushbuttons - I can see what you mean.....

Fixed, as mentioned above :) As you said, it was  odd.

> Toolbar buttons - I have never seen this underline in Communitheme buttons

Believe me, I have it right now -- it's from a few days ago ;) A version from a week ago was like you said and I changed it with an update. More updates on my side will remove such differences.

> View pulse - please, even if you have animation this needs to be fixed, its distracting.

It shows a focus event and I like it ;) The flashes you'd mentioned a log time ago are all removed.

> Editable Combos - really? Just a slight line (not touching the frame) looks ugly?

The line isn't ugly at all; the shadow below the button part is ugly -- while the line-edit part doesn't have such a shadow. If I removed that shadow -- like in the GTK theme -- non-editable combos wouldn't have it either and so, they wouldn't look like buttons.

> As to the scrollbar in combo menus see...

I took the following screenshot while the combo scrollbar was fading out:

![combo_scrollbar](https://user-images.githubusercontent.com/981076/38166004-a172c7f6-3531-11e8-9c36-91eed745c043.png)

However, I've seen that ugly scrollbar from your screenshot in some font combo menus. Is it related to stylesheets? Does Qt draws it itself in the case of some combos? I don't know yet but it isn't related to the current theme. Further investigation will surely be needed.

Anyhow, lots of modifications will come :) I uploaded this only to know about your suggestions.

**Comment by CDrummond** [comment]
Toolbar buttons - as stated I update **every** day. It has **never** been this way. The nearest to a dark toolbar would be the GtkHeader bar. This has buttons as follows:

Hover:
![hover](https://user-images.githubusercontent.com/1141694/38166193-9fed5e9c-3517-11e8-9e30-c019d303a301.png)

Pressed:
![pressed](https://user-images.githubusercontent.com/1141694/38166194-a646cf62-3517-11e8-8a2f-d6a04ddf94b5.png)

**Comment by tsujan** [comment]
Oh, now I understand: I think those headers are more elegant for Qt toollbar buttons, while you think Qt toollbar buttons should be like those GTK buttons. Elegance aside, toggled toolbar buttons wouldn't be recognizable in the second way; would they? While in the first way, they have an orange line below them.

**Comment by CDrummond** [comment]
This comes down to the argument if this is a Qt theme, or a counterpart to the Gtk theme. As per KvAmbiance, I _think_ this should reflect the Gtk theme (wherever possible). This is only my opinion, and you are doing the work, but I **strongly** suggest that KvCommunitheme should try to 1:1 mimic the Gtk counterpart.

I agree that the pressed Communitheme dark buttons are **very** suboptimal. But, then an issue should be filled with Communitheme - so that it can be fixed across Gtk and Qt.

Where you disagree with Communitheme, then this should be under a different name. (as in KvAmbience and and KvAmbiance)

Again, all IM**V**HO

**Comment by CDrummond** [comment]
Pressed button + toggled/pressed == the same (to me). As stated, with the dark GtkHeaderbar they're not that great. (Communitheme is not finished) For the Qt file dialog, I have a light toolbar anyway...

Perhaps, for Kvantum, there should be no dark toolbar? This would resolve the issue. The Gtk theme has light toolbars, its just the headerbar that is dark. There is also talk of using a light menubar (https://github.com/ubuntu/gtk-communitheme/issues/281). If a light menubar is used, then dark toolbars would be odd for Qt apps.

When I said `disagree with Communitheme` - I didn't mean the name (which will change). I meant in the look of widgets.

Bold text in selected tab would be good - and not just for communitheme.

**Comment by tsujan** [comment]
> Perhaps, for Kvantum, there should be no dark toolbar?

No, the dark toolbar is nice -- as you'd suggested -- and it should be alongside a dark menubar, unless they do a bad thing and remove all darkness from the gtk theme.

Menus should have been dark too. Ambiance was much better. Or, maybe, GTK has a new issue with dark menus in light themes?!

> I meant in the look of widgets.

Only some of them. This theme matches Communitheme. Probable differences will be inevitable, either practically or aesthetically.

The problem with making a GTK theme is that you write thousands of lines without a WYSIWYG way of looking at it. Starting a theme months ago and still changing its look shows a deep problem with GTK, not with Ubuntu, IMO.

**Comment by CDrummond** [comment]
Menubars are dark currently, just there is a suggestion to make them white.

**Comment by tsujan** [comment]
I'm just curious and you're familiar with the progress much more than I am: isn't a dark menu suggested? Wouldn't it be elegant?

**Comment by tsujan** [comment]
Took the middle way.

**Comment by CDrummond** [comment]
Looking sooo much better now, thanks for the changes. I agree the pressed dark toggle buttons need making clearer in Communitheme. I'll raise an issue with them.

Another issue: View headers should be bold.

**Comment by tsujan** [comment]
Thanks for checking it again!

The strange thing is that I updated the theme hours ago but:

(1) The active tab doesn't have a bold text;
(2) Headers aren't bold; and
(3) There's no inactive state.

![gtk_theme](https://user-images.githubusercontent.com/981076/38171462-56030280-35af-11e8-8975-16cc28729a38.png)

Maybe I should install Ubuntu on VirtualBox before it gets stable.

**Comment by CDrummond** [comment]
Ah, you're using the PPA? AFAIK, this is no longer updated for 17.10 - the code has moved to be based on 18.04. However, I build from git, and it still builds for 17.10 - hence I **know** I'm always on he latest.

Top window is an inactive window, lower is active:

![screenshot from 2018-04-01 09-54-04](https://user-images.githubusercontent.com/1141694/38171505-cf99ec98-3592-11e8-830d-f95afd9542d6.png)

No need to install VBox, just compile from git - there's not a lot of dependencies, and its fairly trivial.

As to the dark pressed toolbuttons - I've created an issue on Communitheme for this (https://github.com/ubuntu/gtk-communitheme/issues/302) I think it should be solved the same for both Gtk and Kvantum.

**Comment by CDrummond** [comment]
Focus - this should be a solid (rounded) orange border, even round checks/radios. With tabs it should go around the complete 'tab' (ie. the top part) as in on all sides. (Don't particularly like this but that's the way the Gtk variant currently does this for tabs)

**Comment by CDrummond** [comment]
I agree about the tab focus (as stated), but not the dotted focus. IMHO, focus should be consistent - therefore the same style/colour.

But again, this comes down to being consistent with Communitheme. Where you disagree with their direction, I really think you should file issues on their github page - they're a **very** amenable and friendly bunch. Don't forget this is developed by volunteers, not Canonical.

**Comment by tsujan** [comment]
@CDrummond I got the source from https://github.com/Ubuntu/gtk-communitheme/ and did
```
mkdir build && cd build
meson ..
ninja
ninja install
```
The theme is installed correctly and is OK but it's exactly the same. The difference between bold and normal texts:

![bold](https://user-images.githubusercontent.com/981076/38175090-453d6b38-35ec-11e8-95dc-aa3029333d03.png)

From your last screenshot, I see that you use Unity. Can it be the cause of this magic?

I'll add a key to Kvantum for bold text with active tabs -- it's a good feature -- but can't understand this difference....

**Comment by tsujan** [comment]
OK, they aren't bold but `font-weight: 500;` is used for them and that may result in different texts with different fonts. I'll make them bold though. However, the "fallback" mode is still a mystery. I'll make it based on the CSS file.

**Comment by CDrummond** [comment]
I use GNOME shell with Ubuntu fonts.

**Comment by tsujan** [comment]
> Can Kvantum specify font weight?

Not yet. A work for later. For now, bold will do it.

I added an inactive state and made the active tab and header texts bold.

The focus rects are made orange but lighter than Communitheme. The slider focus state is reversed as a compromise because an orange rect around the whole slider was ugly. The focus rects can't be made rounder because of the lack of space.

I also made inactive check/radio buttons and slider indicators dimmer.

**Comment by tsujan** [comment]
I want to implement a key for "boldness percentage" because setting the font weight is wrong (it overrides user's font size). Which font do you use? I want to test with it.

**Comment by CDrummond** [comment]
I use the standard Ubuntu fonts - Ubuntu Regular 11

------

Just updated - and slider focus still looks bad. 

Kvantum:
![qt](https://user-images.githubusercontent.com/1141694/38178028-b90a4d14-3601-11e8-813f-74d01864dca2.png)

Gtk:
![gtk](https://user-images.githubusercontent.com/1141694/38178029-bb1bf742-3601-11e8-80e2-c61306db1f6e.png)

If the filled area is almost the whole slider, then basically nothing is shown for KVantum's focus. (If I look closely I can see the orange around the blue, but its hard to see).

As an aside, sliders with tick marks dont get blue fill. Only sliders without tickmarks - and these have different handles. Why this is, I'm not sure... I've raised a Communitheme issue to ask about this (https://github.com/ubuntu/gtk-communitheme/issues/305) 

Can Kvantum support two different slider handle styles?

---

Don't do this now - as I think Gtk should change(https://github.com/ubuntu/gtk-communitheme/issues/194) - but, is it possible for a Kvantum theme to **not** react to mouse-over on widgets in inactive windows? I think the current behaviour is correct, and Gtk should change to match this, just curious.

---

Do you think the MDI windows should look like the actual windows? e.g. same colour titlebars, and buttons? Not a big issue, as not many apps use these (Qt designer only?)

---

The toolbar/menubar background also needs to change for inactive windows.

![screenshot from 2018-04-01 23-24-02](https://user-images.githubusercontent.com/1141694/38178116-cbacbff4-3603-11e8-87a7-231ebb7434df.png)

**Comment by tsujan** [comment]
> slider focus still looks bad

It isn't bad *at all*; it's just not identical to the gtk theme ;) Long story short, I should take into account various kinds of themes that are possible with Kvantum and changing this for a specific theme will break some others (this kind of focus area can have a semi-transparent interior with some themes).

> As an aside, sliders with tick marks dont get blue fill. Only sliders without tickmarks - and these have different handles. Why this is, I'm not sure... I've raised a Communitheme issue to ask about this 

IMHO, such details aren't important for fair matching. We don't want to (and can't) reproduce the gtk theme in Qt (or conversely).

>  is it possible for a Kvantum theme to not react to mouse-over on widgets in inactive windows?

LOL! I'm sure you don't want me to make Kvantum's behavior wrong because a gtk theme has a wrong behavior ;)

> Can Kvantum support two different slider handle styles?

With and without tick marks? Maybe... should experiment with it later, after more important things are done.

> Don't do this now - as I think Gtk should change

A value of 2 for boldness is good for now (== QFont::DemiBold). It's the same as bold for most fonts but I'll try Ubuntu Regular if I can install it under my Arch-based system. Anyhow, this new feature (for which I'll make a new commit soon) seems good in itself.

> Do you think the MDI windows should look like the actual windows?

Yes, the titlebar should be made dark and buttons need some work. But every WM have its own buttons.

> The toolbar/menubar background also needs to change for inactive windows.

You're completely right. I'll change it later. Thanks!

**Comment by tsujan** [comment]
Succeeded in adding an extra no-tick handle:

![sliders](https://user-images.githubusercontent.com/981076/38184347-3da5de8a-365c-11e8-8363-060998730c13.png)

However, before pushing a commit, I should find a workaround for apps that don't respect the handle size; `QStyle` is something many Qt devs ignore when they see their apps look "good" with Fusion.

**Comment by CDrummond** [comment]
Sliders - IMHO, the fill should be consistent.It _is_ important for consistency. As a user I'd be confused why one app does it, and another does not. But, this is something I think should change in the Gtk variant (hence raising an issue).

As to the mouse-over on inactive windows. Again (IMHO) Gtk **is** wrong (and I have raised an issue). **But** if this is to stay as it is then Kvantum should allow for this. If its a 'feature' that can be mimicked, it should be. But, for now, leave Kvantum's mouse-over as is. It needs the Communitheme issue to be closed, so that a consensus is reached. Perhaps you should add you comments/opinion to that issue?

As a user, I want my desktop to be as consistent as possible. (I've always wanted this, hence QtCurve and KGtk). To me, it should not matter what toolkit an app uses - it should blend in with the current DE. (For instance, I'm really trying to get Cantata to look more GNOME native (and KDE native)). Now I know Gtk apps with GtkHeaderbar don't do this very well, but that's another issue. With themes such as this, then it **is** possible to reduce the differences (and with QGnomePlatform Gtk dialogs are used, etc). So, wherever possible it'd be **fantastic** to match 1:1

Features you disagree with in Communitheme should be raised as issues on its issue tracker - so that they can be resolved in **both** Kvantum and Gtk. Just stating "this is broken, I'm not doing it this way" breaks the uniform desktop. Therefore, implement it your way in Kvantum then raise an issue with a screenshot (if applicable) on Communitheme's bug tracker.

I know that Kvantum's goal is **not** to be a Qt equivalent for Gtk - but in this Communitheme case it **is**

**Comment by CDrummond** [comment]
As for line-edits. Gtk does look nicer with lighter focus on the outside. Can the line-edits not be drawn with a 1 pix transparent border when not focused, allowing for the eternal glow when focused?

**Comment by tsujan** [comment]
> So, wherever possible it'd be fantastic to match 1:1

Maybe, Kvantum sometimes causes an illusion about that because many things that seemed impossible before are now possible with it. However, there's a limit to it. I won't be surprised if someone who has never got his hands dirty in `QStyle` uses the word "wrong" for a dissimilarity between a Qt and a gtk theme but, frankly, I don't understand that from the creator of QtCurve or Cantata.

> "this is broken, I'm not doing it this way" breaks the uniform desktop.

My priority is developing a nice style plugin for Qt. In short, Qt is my priority. You could know about my taste by looking at themes like `KvSimplicityTurquoise`, `KvCurves3d`, `KvGnomish`, `KvFlatLight`, etc. With only 2-3 gtk-related apps, gtk has never been a priority to me and I've never liked gtk3 themes; to me, most of them have elementary looks in spite of thousands of lines written for making them.

That being said, I'm sure you see how HARD I work to make Kvantum themes that match gtk themes. But I never do that by breaking Kvantum or damaging its properties that can be used by purely Qt themes.

BTW, I think I've found a workaround for apps that don't respect slider size.

**Comment by CDrummond** [comment]
I know you work hard, and I know Communitheme is probably not your taste - so thanks for you hard work. As stated multiple times, its more than usable as is now - even with no more changes. In fact, I am currently using KvCommunitheme as my default.

All I was asking was to not deliberately introduce a difference, when it need not be there. I know full well that Gtk is not Qt (and vice versa). They each have widgets/capabilities that the other does not have. (e.g. Kvantum highlights the currently focused view, Gtk does not (and I'm **not** asking for this to be removed)) The line-edit highlight is a difference that I think should be changed. I've stated how I believe it could be accomplished - but if you think KvCommunitheme's look better, then ask Communitheme to change to that (or comment on https://github.com/ubuntu/gtk-communitheme/issues/273)

Communitheme is not perfect, not by a long shot (not insulting it, I could do nowhere near as good) - hence why asking if there are things you think should be done better to raise issues.

Surely, aiming for as consistent as possible a desktop is a good thing? Whether it means changing KvCommunitheme or Gtk Communitheme.

Would it help if I created a check-list (in this issue, or a new issue) where we can discuss the differences of each widget. Even if it was just to state - "this widget is this way becase I like it" (Don't want to endlessly argue over little changes, and this issue is getting rather large!)

And, in **no** way do I want a change made for this theme that will break others - that would be wrong. Just want to see if things can be added - like the different slider types (I originally only asked if it was possible).

**Comment by tsujan** [comment]
OK, I'll skip general statements and explain specific things. As for line-edits:

Combos should align with buttons when put alongside each other horizontally. In Qt, (1) the shadow thickness should be included in the height (and width) and (2) line-edits are drawn in editable combos and spin-boxes exactly as they're drawn alone. So, an outer shadow will break the look of spin-boxes as well as the horizontal alignment of *non-editable* combos with buttons.

**Comment by tsujan** [comment]
One word about corner rounding: It's done by a method I call "frame expansion". With an outer shadow, there wouldn't be enough space for that. I'd considered such things -- and more -- when starting to make the theme.

**Comment by CDrummond** [comment]
Differences to Gtk Communitheme:
- Buttons; pressed shadow is different (nice, but different), but whole look _might_ be changing
- Checks/radios icon should be larger
- Checks/radios in views should be monochrome when selected - e.g. white frame with which dot/check

Not stating that these need to be changed, just pointing out...

Bugs?
- Text of current tab seems lighter/different to that of a groupbox title. Looks like one is bold, and the other is using a different weight. I _think_ all 'bold' texts should use the same weight - whether that is just using `QFont.setBold(true)` or specifying an exact weight doesn't matter, but they should be consistent
- Table headers are missing lines. 1st horizontal header missing left line, and last vertical header missing bottom line. Can these even be fixed? Or would it affect the treeview headers? If so, then leave as is - treeview is used in more places than tableview (I think)

**Comment by tsujan** [comment]
> Buttons; pressed shadow is different (nice, but different), but whole look might be changing

Communittheme's are ugly; KvCommunitheme's are nice. And the difference is nicely subtle.

> Checks/radios icon should be larger

I'd checked before but if you suggest another size, tell me the number of pixels please!

> Checks/radios in views should be monochrome when selected

Tell me the exaxt colors please!

> Bugs?

Bugs? Really?!

> Text of current tab seems lighter/different to that of a groupbox title. Looks like one is bold, and the other is using a different weight. I think all 'bold' texts should use the same weight

I could make their weights equal; a piece of cake.

> Table headers are missing lines. 1st horizontal header missing left line, and last vertical header missing bottom line

You told me to do that :) If those lines are drawn, they will be drawn everywhere --  treeview headers included -- because, in Qt, a header is a header everywhere. Even if I could distinguish between treeview and table headers in the code and introduce a new section for the latter, it might be an overkill. HOWEVER, I'll think about it. If there'a simple and backward-compatible way, why not?

**Comment by CDrummond** [comment]
Buttons - beauty is in the eye of the beholder :-) Anyway, I _knew_ you would say this :-) I don;t like the current Communitheme button look, and raised an issue to change this.
Check/radios of selected items (i.e. with highlight colour) are white border + white dot/check.
Check/radios are 18px x 18px

As to bold font weights - please do make them consistent.

Tableview headers - nah I said for treeviews :-) But, I get tour point - and I did state if it'll break treeviews then not to bother. Can they not have different styles anyway?

**Comment by tsujan** [comment]
> Check/radios of selected items (i.e. with highlight colour) are white border + white dot/check.

Yes, I already know that but adding yet another set of checkboxes for view-items -- there is already an extra set for menus -- seemed like an overkill to me. The code should search the SVG image for extra elements and if I don't stop making new sections, the CPU usage will be out of control. However, I'll look into it and see whether searching could be done only once -- it seems possible.

> Check/radios are 18px x 18px

It looked too big before but I'll test again.

> Tableview headers - nah I said for treeviews :-)

I'd understood your point completely; maybe I used wrong words to explain it. But I'll look into it.

**Comment by tsujan** [comment]
I bravely added optional view-item checkboxes to the code (not pushed yet) and saw no difference in CPU usage (negligible). Apparently, Qt's SVG search is very optimal. I'll add a set to this theme soon.

**Comment by tsujan** [comment]
I think this is perfect:

![item-check](https://user-images.githubusercontent.com/981076/38311397-5b8934ae-3834-11e8-9e64-b6e63b1849bb.png)

Qt's view-items don't have radio buttons. I'll make a commit tomorrow.

BTW, you were right about the size of checkboxes/radiobuttons. However, it shouldn't be greater that `PM_SmallIconSize` (= 16px) in menus and view-items; in other places, it should be 18px with this theme.

**Comment by tsujan** [comment]
Done, both in the code and in the SVG file.

As for tables:

The problem with removing the right/left frame of headers is that they can be "removed" in different ways -- for example, they can be transparent or have the view background color. Moreover, Kvantum gets the drawing from the SVG file and the frame image can be very simple, as in this case, or complex, as in some other themes. So, in the code, there's no way to know whether a frame is removed. (There's also another problem, which is an old Qt bug and shows up with such settings in rare cases.)

So, there are 2 ways: (1) leaving it as it is; or (2) adding a margin between the header and the view border, like in `KvAdapta`. I knew that you wouldn't agree with (2), while (1) wasn't so bad to me; so, I didn't touch that part of the SVG file.

The margins mentioned at https://github.com/tsujan/Kvantum/pull/192#issuecomment-378511747 are also made equal to each other.

**Comment by tsujan** [comment]
I think the theme is ready. I'll make changes to it whenever the GTK theme changes. The dark variant should be also made but that should be much easier.

**Comment by tsujan** [comment]
Yes, they're darker in the gtk theme. I'll change them later.

**Comment by CDrummond** [comment]
Not sure why this issue is closed, when Communitheme is not finished? Should it not be left open?

I know your feelings on being a 1:1 clone, but a few observations of differences which might be nice to remove:

1. Tab focus is now just a dotted rectangle within the tab - I do like Kvantum's, but favour consistency more.
2. Push buttons have less shadow when raised, and much less of an internal shadow when pressed
3. Toggled toobar items are different. Gtk has dark border, lighter background, 1 pix inner shadow at top (as per regular buttons)
4. Non-tick mark slider handles are circular
5. Checked check/radios in list/tree views remain green

Looking at KvantumPreview with KvCommunitheme, there is a **lot** of black (well not black, but dark grey) at the top. What about only drawing the primary toolbar black if there is no menubar?

Perhaps decrease the height of the menubar? (Is that even configurable with Kv styles?)

Also, the menuitem text of combo menus is too light.

**Comment by tsujan** [comment]
I closed it because, apparently, no final decision is going to be made about the GTK+ theme. There are too many developers and each one has a different opinion.

Thanks a lot for the info! I'll implement the changes you mentioned whenever I find some free time (being VERY busy with several projects, mine and others'). Please keep telling me about such changes!

**Comment by CDrummond** [comment]
That is fair enough! Perhaps, what I should do is wait a little while longer - until Communitheme has settled more, and then make a list of differences. You can then decide to change KvCommunitheme to match or not on a per-difference basis. 

There is no point you wasting your time making a change, which is then reverted or changed to something completely different. (To be fair, Communitheme  does seem to be settling, most changes are to fix bugs - its name is going to be changed as well).

**Comment by Feichtmeier** [comment]
Hi @tsujan !
Thank you for this communitheme Qt theme.
The flood of changes and complicated design switches has settled a bit :smile: 
I am also working on the dark theme which you can see here: https://github.com/ubuntu/gtk-communitheme/pull/444
And we are about the enter the phase of refinements and bug fixes.
Let us know if you have any questions or ideas please!
Best wishes

**Comment by tsujan** [comment]
@Feichtmeier 

Thanks for the good news!

I'll wait a little more and then try to reflect the changes in the Kvantum theme as far as possible. I'll also check your theme once in a while to know about probable changes.

**Comment by Feichtmeier** [comment]
Hi @tsujan =)
If we want to include your awesome work into the communitheme: could you provide some basic information on how I can test it at the moment? 
Since I work from my  host machine on the theme I use the meson installation from our repo and not the snap.
Later on - if we tested your theme in a non snap environment successfully - we should include it in the snap installation.

I lack some basic knowledge about qt theming and your repo: so your work is a theme engine and the communitheme adaption is "only" 1 theme in that engine?

Best regards

**Comment by tsujan** [comment]
@Feichtmeier 

Kvantum is a Qt style engine with a default (E-like) theme. It comes with many (extra) themes, to which `KvCommunitheme` was added recently. Other themes are also made for Kvantum; they can be installed and activated by using `Kvantum Manager`. Each them consists of a `THEME.kvconfig` file and a `THEME.svg` image and can be installed in various places, including `/usr/share/themes/THEME/Kvantum/` for more compatibility with GTK themes. Their total size is ~157 KiB for `KvCommunitheme`.

Kvantum isn't made to imitate GTK+ themes but the fact is that it can do so to a great extent -- at least, much better than other Qt style engines. There will are always inevitable differences but the general looks and feels can be made very similar.

I think it's better for `KvCommunitheme`  to be shipped with Kvantum, as it is now, because I'll change it once in while, following the Ubuntu theme or optimizing it. Of course, other themes can also be made based on it.

Recently, I've been very busy but I haven't forgotten about `KvCommunitheme` at all. I'll try to update it soon this week.

BTW, I add the prefix `Kv` to the names of themes I add to Kvantum for them to be distinguished from the themes made by others, inclusing GTK themes. It isn't necessary though.

**Comment by CDrummond** [comment]
@Feichtmeier at the risk of sounding like a broken record... You can use [QGnomePlatform](https://github.com/MartinBriza/QGnomePlatform) to automatically set the Qt theme based upon the Gtk theme. This also works for Kvantum themes. QGnomePlatform reads the user's gtk theme setting, and looks for a matching Qt theme (e.g. Adwaita-Qt) or for a matching Kvantum theme. This works both for the `Kv` themes that Kvantum ships with, and Kvantum themes installed into (e.g.) `/usr/share/themes`

This platform also allows Qt apps to use Gtk file, color, and font, dialogs.

To use this, you just need to set the `QT_QPA_PLATFORMTHEME` environment variable to `gnome` - e.g. QT_QPA_PLATFORMTHEME=gnome

**Comment by tsujan** [comment]
@CDrummond Thanks! I didn't say anything about `QGnomePlatform` because I thought everyone knew about it under GTK DEs. Apparently, I was wrong.

**Comment by tsujan** [comment]
> So if it will be included in 18.10, could the kvantum engine + the kvcommunitheme files be included into the snap?

If you change "KvCommunitheme" to "Communitheme" (just the name, everywhere), the files could be put inside `/usr/share/themes/Communitheme/Kvantum`. Then, if Kvantum is installed, it'll see the theme; If not, nothing bad will happen. Moreover, Kvantum supports global kvconfig file(s) `$XDG_CONFIG_DIRS/Kvantum/kvantum.kvconfig` as follows, so that distro themes could be loaded by default, instead of the default Kvantum theme:
```
[General]
theme=NAME_OF_THEME
```
More tweaks can be done too (for example, different themes for different Qt apps) but, usually, they're done by users easily (with Kvantum manager).

> So we would need all - QGnomePlatform, kvantum engine and kvcommunitheme - to be included into 18.10 somehow to theme Qt apps with the communitheme?

Yes! Although `KvCommunitheme` will come with Kvantum automatically, you might like to have your own version of it as described above.

> @tsujan the name is not so important at the moment

I agree. "Communitheme" sounded a little weird at first but I've got used to it now :)

**Comment by tsujan** [comment]
I changed buttons today but will make a commit after all changes are made (which may take a while):

![buttons](https://user-images.githubusercontent.com/981076/40388658-55014fa6-5e25-11e8-9850-b0d6c3ab0ae7.png)

**Comment by tsujan** [comment]
Made many changes to KvCommunitheme: https://github.com/tsujan/Kvantum/commit/f76d6fbe66a7ce74427e21dfcbdf5f0f5c045d19

@CDrummond

(1) For now, I left the tab focus as it was -- for the sake of elegance. Anyhow, Qt tabs are different from GTK ones.

(2) As for the dark toolbars, their height can be made smaller but (a) many apps (including Cantata) enforce it; (b) 22px wouldn't make any difference and GTK doesn't use 16px toolbar icons; and (c) they're elegant, IMO.

Other things you mentioned are considered in the commit. I also made the combo-boxes more similar to GTK combos but, of course, a 100% similarity isn't possible. Other changes are made too.

**Comment by CDrummond** [comment]
Just updated - and looks GREAT! Some notes...
1. Pressed toggled buttons have no mouse over effect. Gtk has same issue https://github.com/ubuntu/gtk-communitheme/issues/457
2. Any chance of rounded menus?
3. Default buttons should have a darker border, and not bold text.
4. Can you theme the MDI controls to match Communitheme's window controls?
5. Tooltips have a glossy like effect. This does not match the rest of the theme, and they should be flat.

**Comment by tsujan** [comment]
> Pressed toggled buttons have no mouse over effect

In Kvantum, they never have.

> Any chance of rounded menus?

Yes, of course, but are GTK menus rounded? In Kvantum, if menus are rounded, there should be an extra space for them ("frame expansion" doesn't apply to them), so that menu items won't cover the whole width.

> Default buttons should have a darker border, and not bold text.

I didn't know that; will check.

> Can you theme the MDI controls to match Communitheme's window controls?

OK, later.

> Tooltips have a glossy like effect.

That's because I didn't include a tooltip (to have a lighter theme); so, the default Kvantum tooltip is used. But OK, I'll add it.

Thanks!

**Comment by tsujan** [comment]
I checked and saw that gtk menus are just a little rounded at the bottom. That isn't possible without an extra space with Kvantum. Actually, many Kvantum themes have rounded menus but their menu items don't cover their whole widths (or heights). I'll see if I could generalize "frame expansion" to menus but no promise.

**Comment by Feichtmeier** [comment]
Hello @tsujan and @CDrummond 
Do you have accounts on the Ubuntu community hub?
I think it would be better if you would talk to someone who has even slightly more knowledge of this :)
So we could make Qt theme thread in there and discuss what's on the todo list to make this part of the new ubntu theme, if you still want to, ofc :smile: 
Things are going into the hot phase soon, so it would be great if we could make it in time.

**Comment by tsujan** [comment]
@Feichtmeier I'm not an Ubuntu user but @CDrummond is. He's also very knowledgeable in this area -- actually, I've adapted some of his codes in Kvantum and he's added gtk-related codes to it.

**Comment by CDrummond** [comment]
@Feichtmeier I have mentioned the required projects, etc, multiple times - on both didrock's blog, the [community forum](https://community.ubuntu.com/t/qt-integration-for-18-04/4323), and even [Canonicals github bug tracker](https://github.com/CanonicalLtd/desktop-design/issues/65). So, Ubuntu/Canonical know what needs to be done. Not sure what more I can realistically do?

**Comment by CDrummond** [comment]
@tsujan some **very** minor nit-picky issues....

1. Menu separators are not drawn. This is fine, and matches Gtk. But, Kvantum seems to reserve a lot of space for them - more than Gtk
2. The dark border for pressed toolbar buttons on a dark toolbar is too thick. The Gtk variant only has a 1 pixel border, but draws the top darker.
3. Restore icon in MDI buttons is incorrect. It should be a hollow square, with a slightly lighter square behind this (where you can only see the top and side of this other square)
4. The Gtk variant currently uses orange for most highlights, but a light blue for highlighting editable text. (To be honest, I'm not sure I like this - but it'd be nice to be consistent)
5. List/tree headers should be the same colour as main widget background
6. Pressed buttons have a darker border on the left than right - only the top should be darker. (As in for the 3d effect, the light is coming from the top, and not to one side).

**Comment by tsujan** [comment]
Thanks again for your attention to details!

1. The space for menu separators is fixed. I've had no plan to make it variable but will see if it's worth adding a parameter for it.
2. OK.
3. gtk3-widget-factory doesn't show a restore icon. Where can I find it?
4. What do you men by "highlight"? Hover highlight? Selection highlight?
5. I made them like in the version I had. Are they changed?
6. Like 5.

**Comment by CDrummond** [comment]
1. But, with there being no line, the empty space looks quite odd.
2.
3. Screenshot:
![titlebar](https://user-images.githubusercontent.com/1141694/40787276-20c3a2b8-64e5-11e8-8af1-df37006dc3df.png)
4. Ah sorry, the selection colour - or `QPalette::Highlight` e.g. select an area of text in a line-edit. See https://github.com/ubuntu/gtk-communitheme/pull/497
5. I guess so
6. AFAIR, the shadows have only ever been from above. 

As stated, I am being **really** picky here - non of this is anything major.

**Comment by tsujan** [comment]
The restore button is for maximized, shaded and minimized windows. Now I see that GTK has it for maximized windows. I can make it like that.

If the highlight is always blue, I'll make it blue but if it's sometimes blue, sometimes orange, that won't be possible with Qt.

> As stated, I am being really picky here

And that's good :)

**Comment by CDrummond** [comment]
Yeah, as stated, I'm not sure I like 2 different highlight colours. I guess a work-around would be to change the palettes of QLineEdit and QTextEdit so that they use blue. But, to do this nicely would mean adding custom palettes for all widget types - and that is probably more hassle than its worth.

**Comment by tsujan** [comment]
Done in https://github.com/tsujan/Kvantum/commit/489770a2c72c1672cf2030845b18f91f4c539aee, except for the restore button and header bg color, for which I should update the gtk theme on another computer with gnome-shell.

2 or more highlight colors might be possible but wouldn't be worth the effort and its code wouldn't be maintainable.

**Comment by NikoKrause** [comment]
I'm seeing a flickering using the KvCommunitheme. (This flickering doesn't happen with other kvantum themes).
I use Mint 19 Cinnamon with Kvantum. When changing the sound volume the window of QT apps (Okular, Kvantum Manager, etc) flickers. This also happens when dragging a window.
This can be seen in the following screenshots. A normal window looks like this:
![screenshot-window-2018-06-05-233519](https://user-images.githubusercontent.com/8415124/41004362-e80243a6-6919-11e8-97f9-c5ce1ffa99d8.png)
And while dragging the window it looks like this:
![screenshot-window-2018-06-05-233531](https://user-images.githubusercontent.com/8415124/41004386-f953d98a-6919-11e8-864d-b1d449ab5bbb.png)

**Comment by tsujan** [comment]
@NikoKrause 
That's a known problem GTK window managers have wit Qt apps (or, conversely, a problem Qt has with GTK DEs): They make Qt windows *inactive* with dragging. `KvCommunitheme` has an inactive state (only Kvantum supports inactiveness, until now), so you could see that effect with it. Some other Kvantum themes have inactive states too; try `KvSimplicityTurquoise` or `KvAmbience`, for example, under a GTK WM.

It doesn't happen under KWin.

**Comment by tsujan** [comment]
BTW, if you don't drag a Qt window and just focus another window, the first one will be "inactive" and that state will be shown by `KvCommunitheme`. It isn't a bug but a feature. GTK themes also support the inactive state but they call it the "backdrop" state.

**Comment by NikoKrause** [comment]
P.S.: Communitheme looks pretty nice. The inactive stat while dragging doesn't bother me. What bothers me is the flickering when changing the sound volume. (But since I'm using Arc and it doesn't happen there, I'm om with it ;)

**Comment by tsujan** [comment]
It may be a good idea to add a key to Kvantum Manager for ignoring the inactive state in themes that have it. I'll implement it if it doesn't need many changes.

**Comment by tsujan** [comment]
I added a checkbox for ignoring the inactive state to Kvantum Manager → Compositing & General Look. It works with themes that distinguish between active and inactive states in any way.

**Comment by tsujan** [comment]
Changed the header color too.

The restore button isn't just for restoring from the maximized state. Communitheme's restore button can't be used here.

I think KvCommunitheme is a fair clone of Communitheme now and this issue can't remain open indefinitely. Please report future problems as separate issues!

**Comment by VandaLHJ** [comment]
Is there a chance to get windows decorations to match kvcommunitheme?

**Comment by tsujan** [comment]
> Is there a chance to get windows decorations to match kvcommunitheme? I'm using KDE plasma.

Kvantum is a Qt style engine, not a window decoration. Under KDE, the decoration is provided by KWin by default. The breeze decoration is ugly with this theme, unless you change its color by editing the color scheme but an Aurorae decoration can be made.

**Comment by tsujan** [comment]
A better way is making a special color scheme and assigning it to all title-bars without applying it to widgets. KDE can do it.

**Comment by VandaLHJ** [comment]
I know that Kvantum don't provide windows decoration, i currently use Materia KDE (Aurorae) for both top bar and windows decorations. It work kinda ok, but isn't perfect match, but best i was able to figure out without any tweaks.

**Comment by tsujan** [comment]
You could make an appropriate color scheme that gives a dark brown color to the titlebar (window background) and use it with Breeze by going to System Settings -> Window Management -> Window Rules -> New -> Appearance & Fixes -> Titlebar color scheme -> Force.

Breeze titlebar is flexible and can also assign different color schemes to titlebars of different apps. That goes very well with Kvantum, which can assign separate themes to different apps.
